### PR TITLE
[SPARK-44071] Define and use Unresolved[Leaf|Unary]Node traits.

### PR DESCRIPTION
**Pull request by ryan-johnson-databricks**
### What changes were proposed in this pull request?

Looking at [unresolved.scala](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala), catalyst would benefit from an `UnresolvedNode` trait that various `UnresolvedFoo` classes could inherit:
```scala
trait UnresolvedNode extends LogicalPlan {
  override def output: Seq[Attribute] = Nil
  override lazy val resolved = false
}
```
Today, the code is duplicated in ~20 locations (7 of them in that one file).

### Why are the changes needed?

Reduces redundancy, improves readability, documents programmer intent better.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Mild refactor, existing unit tests suffice.

### CODE REVIEW

**Review by cloud-fan — APPROVED** [review]
LGTM if all tests pass

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

**Comment by ryan-johnson-databricks** [comment]
All green, except the [pyspark-sql build](https://github.com/ryan-johnson-databricks/spark/actions/runs/5282492561/jobs/9557510341) failed with what appears to be upstream breakage?
```
[error] org.apache.maven.model.building.ModelBuildingException: 1 problem was encountered while building the effective model for org.apache.spark:spark-parent_2.12:3.5.0-SNAPSHOT
[error] [FATAL] Non-resolvable parent POM for org.apache.spark:spark-parent_2.12:3.5.0-SNAPSHOT: Could not transfer artifact org.apache:apache:pom:18 from/to central (http://repo.maven.apache.org/maven2): status code: 501, reason phrase: HTTPS Required (501) and 'parent.relativePath' points at wrong local POM @ line 22, column 11
```
(I retried the job but it immediately failed again with the same error -- so probably not an infra issue)

**Comment by MaxGekk** [comment]
@ryan-johnson-databricks Could you re-trigger GitHub actions, please, via an empty commit:
```
$ git commit --allow-empty -m "Trigger build"
```

**Comment by ryan-johnson-databricks** [comment]
@MaxGekk all green!

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @ryan-johnson-databricks and @cloud-fan for review.

**Pull request closed without merging**
